core/src/crds_gossip_pull.rs

```diff
@@ -537,24 +537,21 @@ impl CrdsGossipPull {
     /// The value_hash of an active item is put into self.purged_values queue
     pub fn purge_active(
         &mut self,
+        thread_pool: &ThreadPool,
         crds: &mut Crds,
         now: u64,
         timeouts: &HashMap<Pubkey, u64>,
     ) -> usize {
-        let old = crds.find_old_labels(now, timeouts);
-        let mut purged: VecDeque<_> = old
-            .iter()
-            .filter_map(|label| {
-                let rv = crds
-                    .lookup_versioned(label)
-                    .map(|val| (val.value_hash, val.local_timestamp));
-                crds.remove(label);
-                rv
-            })
-            .collect();
-        let ret = purged.len();
-        self.purged_values.append(&mut purged);
-        ret
+        let num_purged_values = self.purged_values.len();
+        self.purged_values.extend(
+            crds.find_old_labels(thread_pool, now, timeouts)
+                .into_iter()
+                .filter_map(|label| {
+                    let val = crds.remove(&label)?;
```

Comment:
is this `remove() `guaranteed to return `Some` because no other thread is touching the table (I think `purge` is happening under the gossip write lock). If so can we unwrap here()?

Reply:
You are right, but there is nothing locally here that guarantees this is the case and will stay so in the future. For example, we may later in the future want to switch to a concurrent hash-map which does not require explicit locking. But then someone should remember that the code here was relying on single threaded access to the crds table. So I think it would be safer and more forward compatible if we do not rely on that assumption here.

Reply:
@behzadnouri hehe I was actually thinking along the same scenario, i.e. where we switch to a concurrent hash-map. I wondering if it would be better to add an assertion here in that case as well, so that we remember to enforce the guarantee  that with respect to a particular label,`check timestamp->remove` is an atomic operation even in a concurrent hashmap implementation. Otherwise, for instance, an insertion might add a newer version of that label and then we remove the newer label by accident.

Reply:
That is a fair point, and might be the right direction to go. But my suggestion is that, if we want to make that change, lets do it separately in its own commit so that this commit implements [the same logic as the existing code](https://github.com/solana-labs/solana/blob/368aeb2ce/core/src/crds_gossip_pull.rs#L547-L553), but only in parallel.

Reply:
Another thing is that `crds.remove(&label) == None` is not necessarily an error. For example hypothetically if crds table was thread-safe, then 2 threads could be simultaneously doing the same thing; then either one which removes the old value from the table and store it in `purged_values` has already done the job, and the other one can skip.

Reply:
makes sense to me!